vmm/src/cpu.rs

```diff
@@ -778,7 +778,11 @@ impl CpuManager {
         }
 
         // Only create vCPUs in excess of all the allocated vCPUs.
-        for cpu_id in self.vcpus.len() as u8..desired_vcpus {
+        #[cfg(target_arch = "aarch64")]
+        let possible_cpus = self.config.max_vcpus;
```

Comment:
I don't think this is the right design here. So correct me if I'm wrong, but it seems like you're going to create all the vCPU threads no matter if the CPUs are actually plugged into the VM or not. That feels wrong since you will end up with 100 threads running if you start a VM with `--cpus boot=1,max=100`, even if you're never hotplugging any additional vCPU.
I think the implementation between `aarch64` and `x86_64` should not differ here.

Reply:
/cc @rbradford

Reply:
It may not a good way to create all of the max Vcpu at init time. But this is the limitation of armv8/kvm. This issue has been discussed inside and outside Arm for several times. Quotes Sally's Words here:

> B. Limitations of GIC to Support Virtual CPU Hotplug:
1. GIC requires various resources(related to GICR/redistributor, GICC/cpu
   interface etc) like memory regions to be fixed at the VM init time and these
   could not be changed later on after VM has inited.
2. Associations between GICC(GIC cpu interface) and vcpu get fixed at the VM
   init time and GIC does not allows to change this association once GIC has
   initialized.

C. Known Limitation of the KVM:
1. As of now KVM allows to create VCPUs but does not allows to delete the
   already created vcpus. QEMU already provides an interface to manage created
   vcpus at KVM level and then to re-use them.
2. Inconsistency in interpretation of the MPIDR generated by KVM for vcpus
   vis-a-vis SMT/threads. This does not looks to be compliant to the MPIDR
   format(SMT is present) as mentioned in the ARMv8 spec. (Please correct my
   understanding if I am wrong here?)
   
Workaround to the problems mentioned in Section B & C1:
1. We pre-size the GIC with possible vcpus at VM init time
2. Pre-create all possible vcpus at KVM and associate them with GICC 
3. Park the unplugged vcpus (similar to x86)

See the full info [here](https://yhbt.net/lore/all/0126fb3d008d4690950cf843d6ece70b@huawei.com/t/)

Reply:
Ok thanks for the background on this topic :+1: